### PR TITLE
feat: metrics on health checks

### PR DESCRIPTION
**Pull request by brianluong**
# Description

Metrics for health checks — counts, errors, latencies, check results

## Type of change

- [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
- [x] 😎 New feature (non-breaking change which adds functionality)
- [ ] ⁉️ Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] ⚒️ Refactor (no functional changes)
- [ ] 📖 Documentation (updating or adding docs)

# How Has This Been Tested?

```
# HELP node_gateway_healthcheck_block_height Block height of upstream.
# TYPE node_gateway_healthcheck_block_height gauge
node_gateway_healthcheck_block_height{endpoint_id="satsuma-erigon-1",url="http://54.146.12.232:8545"} 1.5454965e+07
node_gateway_healthcheck_block_height{endpoint_id="satsuma-erigon-2",url="http://35.171.183.44:8545"} 1.5454965e+07
# HELP node_gateway_healthcheck_block_height_duration_seconds Latency of block height requests.
# TYPE node_gateway_healthcheck_block_height_duration_seconds histogram
node_gateway_healthcheck_block_height_duration_seconds_bucket{endpoint_id="satsuma-erigon-1",url="http://54.146.12.232:8545",le="0.1"} 5
node_gateway_healthcheck_block_height_duration_seconds_sum{endpoint_id="satsuma-erigon-1",url="http://54.146.12.232:8545"} 0.45845249899999996
node_gateway_healthcheck_block_height_duration_seconds_count{endpoint_id="satsuma-erigon-1",url="http://54.146.12.232:8545"} 5
...
...
# HELP node_gateway_healthcheck_block_height_total Total block height requests made.
# TYPE node_gateway_healthcheck_block_height_total counter
node_gateway_healthcheck_block_height_total{endpoint_id="satsuma-erigon-1",url="http://54.146.12.232:8545"} 5
node_gateway_healthcheck_block_height_total{endpoint_id="satsuma-erigon-2",url="http://35.171.183.44:8545"} 5





# HELP node_gateway_healthcheck_peer_count Block height of upstream.
# TYPE node_gateway_healthcheck_peer_count gauge
node_gateway_healthcheck_peer_count{endpoint_id="satsuma-erigon-1",url="http://54.146.12.232:8545"} 33
node_gateway_healthcheck_peer_count{endpoint_id="satsuma-erigon-2",url="http://35.171.183.44:8545"} 33
# HELP node_gateway_healthcheck_peer_count_duration_seconds Latency of peer count requests.
# TYPE node_gateway_healthcheck_peer_count_duration_seconds histogram
node_gateway_healthcheck_peer_count_duration_seconds_bucket{endpoint_id="satsuma-erigon-1",url="http://54.146.12.232:8545",le="0.1"} 1
node_gateway_healthcheck_peer_count_duration_seconds_sum{endpoint_id="satsuma-erigon-1",url="http://54.146.12.232:8545"} 0.972279084
node_gateway_healthcheck_peer_count_duration_seconds_count{endpoint_id="satsuma-erigon-2",url="http://35.171.183.44:8545"} 6
# HELP node_gateway_healthcheck_peer_count_total Total peer count requests made.
# TYPE node_gateway_healthcheck_peer_count_total counter
node_gateway_healthcheck_peer_count_total{endpoint_id="satsuma-erigon-1",url="http://54.146.12.232:8545"} 6
node_gateway_healthcheck_peer_count_total{endpoint_id="satsuma-erigon-2",url="http://35.171.183.44:8545"} 6







# HELP node_gateway_healthcheck_sync_status Sync status of upstream.
# TYPE node_gateway_healthcheck_sync_status gauge
node_gateway_healthcheck_sync_status{endpoint_id="satsuma-erigon-1",url="http://54.146.12.232:8545"} 0
node_gateway_healthcheck_sync_status{endpoint_id="satsuma-erigon-2",url="http://35.171.183.44:8545"} 0
# HELP node_gateway_healthcheck_sync_status_duration_seconds Latency of sync status requests.
# TYPE node_gateway_healthcheck_sync_status_duration_seconds histogram
node_gateway_healthcheck_sync_status_duration_seconds_bucket{endpoint_id="satsuma-erigon-1",url="http://54.146.12.232:8545",le="0.1"} 4

node_gateway_healthcheck_sync_status_duration_seconds_sum{endpoint_id="satsuma-erigon-1",url="http://54.146.12.232:8545"} 0.7080501670000001
node_gateway_healthcheck_sync_status_duration_seconds_count{endpoint_id="satsuma-erigon-1",url="http://54.146.12.232:8545"} 6

# HELP node_gateway_healthcheck_sync_status_total Total sync status requests made.
# TYPE node_gateway_healthcheck_sync_status_total counter
node_gateway_healthcheck_sync_status_total{endpoint_id="satsuma-erigon-1",url="http://54.146.12.232:8545"} 6
node_gateway_healthcheck_sync_status_total{endpoint_id="satsuma-erigon-2",url="http://35.171.183.44:8545"} 6
```

### CODE REVIEW

File: internal/checks/blockheight.go [hunk 1]
```diff
@@ -122,13 +133,15 @@ func (c *BlockHeightCheck) GetError() error {
...
 		c.blockHeight = header.Number.Uint64()
+		metrics.BlockHeight.WithLabelValues(c.upstreamConfig.ID, c.upstreamConfig.HTTPURL).Set(float64(c.blockHeight))
 		c.webSocketError = nil
 		c.blockHeightError = nil
 	}
 
 	onError := func(failure string) {
 		zap.L().Error("Encountered error in NewHead Websockets subscription.", zap.Any("upstreamID", c.upstreamConfig.ID), zap.String("WSURL", c.upstreamConfig.WSURL))
 
+		metrics.BlockHeightErrors.WithLabelValues(c.upstreamConfig.ID, c.upstreamConfig.HTTPURL).Inc()
```
Comment (dianwen) [inline internal/checks/blockheight.go:144]: Would it be helpful to label the errors? There are many different kinds for block height checking and some are more meaningful than others.

File: internal/metrics/metrics.go [hunk 2]
```diff
@@ -73,13 +76,139 @@ var (
...
+		prometheus.GaugeOpts{
+			Namespace: metricsNamespace,
+			Subsystem: "healthcheck",
+			Name:      "block_height",
+			Help:      "Block height of upstream.",
+		},
+		[]string{"endpoint_id", "url"},
+	)
+
+	BlockHeightTotalRequests = promauto.NewCounterVec(
```
Comment (dianwen) [inline internal/metrics/metrics.go:99]: Would it be more consistent to call this `BlockHeightTotalChecks`? Same w/ the other ones that have `Request` as a suffix.

File: internal/metrics/metrics.go [hunk 3]
```diff
@@ -73,13 +76,139 @@ var (
...
+		prometheus.CounterOpts{
+			Namespace: metricsNamespace,
+			Subsystem: "healthcheck",
+			Name:      "block_height_total",
+			Help:      "Total block height requests made.",
+		},
+		[]string{"endpoint_id", "url"},
+	)
+
+	BlockHeightDuration = promauto.NewHistogramVec(
```
Comment (dianwen) [inline internal/metrics/metrics.go:109]: Nit: `BlockHeightCheckDuration`? Same with the other duration names

File: internal/metrics/metrics.go [hunk 4]
```diff
@@ -73,13 +76,139 @@ var (
...
+			Help:      "Block height of upstream.",
+		},
+		[]string{"endpoint_id", "url"},
+	)
+
+	BlockHeightTotalRequests = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: metricsNamespace,
+			Subsystem: "healthcheck",
+			Name:      "block_height_total",
```
Comment (dianwen) [inline internal/metrics/metrics.go:103]: should we include `requests`, like `block_height_requests` to be consistent with the metrics in the subgraph gateway? I also removed `total` from the end. Same w/ the other metrics.

File: internal/metrics/metrics.go [hunk 5]
```diff
@@ -73,13 +76,139 @@ var (
...
+			Buckets:   []float64{.005, .01, .025, .05, .1, .25, .5, 1, 2.5, 5, 10, 20, 40},
+		},
+		[]string{"endpoint_id", "url"},
+	)
+
+	BlockHeightErrors = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: metricsNamespace,
+			Subsystem: "healthcheck",
+			Name:      "block_height_errors",
```
Comment (dianwen) [inline internal/metrics/metrics.go:124]: `block_height_check_errors`? Same w/ the other errors

File: internal/checks/blockheight.go (hunk 1 above)
Comment (brianluong) [reply internal/checks/blockheight.go:144]: yeah, look at the interfaces it looks tricky to do that atm.

The error returned is a generic `error` type - the only type of information that error provides is an error string - can you put those as labels?

File: internal/metrics/metrics.go (hunk 2 above)
Comment (brianluong) [reply internal/metrics/metrics.go:99]: 👍

File: internal/checks/blockheight.go (hunk 1 above)
Comment (brianluong) [reply internal/checks/blockheight.go]: In our logic I suppose the types of failures are WS ones — initialization failure is one, and the other is an error encountered after successful init. And then there's HTTP failure

Comment (dianwen) [reply internal/checks/blockheight.go]: Errors strings as labels aren't ideal b/c they can be very high cardinality and hard to build graphs off of. I was thinking we could have string enums for categories of errors, such as `get_client_failure`, `http_get_Head_failure`, `ws_subscribe_new_head_failure`, etc and not break down errors within each category.

File: internal/metrics/metrics.go [hunk 6]
```diff
@@ -31,29 +37,31 @@ var (
...
 	)
 
+	// Upstream routing metrics
+
 	UpstreamRPCRequestsTotal = promauto.NewCounterVec(
 		prometheus.CounterOpts{
 			Namespace: metricsNamespace,
-			Subsystem: metricsSubsystem,
+			Subsystem: "router",
 			Name:      "upstream_rpc_requests_total",
```
Comment (dianwen) [inline internal/metrics/metrics.go:65]: Should we remove the `total` suffix?

File: internal/checks/blockheight.go (hunk 1 above)
Comment (brianluong) [reply internal/checks/blockheight.go]: gotcha! made some changes

File: internal/metrics/metrics.go (hunk 6 above)
Comment (brianluong) [reply internal/metrics/metrics.go]: Removed!

File: internal/metrics/metrics.go [hunk 7]
```diff
@@ -12,17 +12,24 @@ import (
 
 const (
 	DefaultPort              = 9090
-	metricsNamespace         = "satsuma"
-	metricsSubsystem         = "node_gateway"
+	metricsNamespace         = "node_gateway"
 	defaultReadHeaderTimeout = 10 * time.Second
+
+	// Metric labels
```
Comment (dianwen) [inline internal/metrics/metrics.go:18]: Might be helpful to somehow namespace these as blockHeightCheckErrorTypes

**Review by dianwen — APPROVED** [review]

File: internal/metrics/metrics.go (hunk 7 above)
Comment (brianluong) [reply internal/metrics/metrics.go:18]: `HTTPInit` and `HTTPRequest` apply to all the checks, where only the WS ones apply to block height. I'll add more comments to clarify their usage.

**Pull request merged**